### ISSUE TITLE
NVMe: save/restore correctness harness at the real-server seam

### ISSUE THREAD

**Issue by dymoo** [dymoo/mimo-halo-lab#63]
## Goal

Build the save/restore correctness harness at the real-server seam: uninterrupted hot execution versus save → erase/free → restore → continue, with deterministic tokens/logits where accessible, physical prompt-token counters and sequence positions. Tiny-model CI must catch full-prefix recomputation, not merely HTTP 200.

## Inputs

- Real server fixture needs from `agent://NvmeTestModel` (JSON-encoded payload): tiny presets `ggml-org/test-model-stories260K`, `ggml-org/stories15M_MOE`, `ggml-org/tinygemma3-GGUF`, zero-download `tests/test-llama-archs.cpp` generator (mimo2-moe/gemma3-dense, token-array prompts); harness `tools/server/tests/{utils.py,conftest.py,tests.sh}`; greedy `temperature 0 seed 42`.
- Physical counters from `agent://NvmeServerAudit`: `timings.prompt_n` (=`n_prompt_processed`), `timings.cache_n`, Prometheus `llamacpp:prompt_tokens_total` / `prompt_tokens_cached_total`, `GET /slots` processed/cache fields.
- Prior art: `upstreams/strix-llama.cpp/tools/server/tests/unit/test_slot_save.py` (prompt_n==21/6/1 assertions; line 217 explicitly skips default-SWA prefix reuse), `unit/test_metrics.py` (processed vs cached split), `unit/test_kv_keep_only_active.py` (RAM-tier eviction precedent).

## Outputs

- New real-server test file(s) under `upstreams/strix-llama.cpp/tools/server/tests/unit/` covering: uninterrupted vs save→free→restore→continue equivalence; physical-prefill assertions; hot-memory-release verification.
- The missing restored-slot + default-SWA prefix-reuse assertion (today's skipped case) present and enforced.

## Blockers

- Blocked by: #60 (slot-serialization audit).

## Acceptance

- [ ] Harness runs on the Mac against a tiny fixture; no hardware required.
- [ ] Output equivalence: greedy deterministic tokens match uninterrupted vs restored runs.
- [ ] Physical prefill asserted from `prompt_n` AND independently from `llamacpp:prompt_tokens_total` delta — a full-prefix recomputation fails the test even when HTTP 200 is returned.
- [ ] Hot memory release verified (slot IDLE and KV actually freed), not merely relabeled.
- [ ] Existing skip for default-SWA prefix reuse replaced by a real assertion or a tracked failing-first regression.

## Commands

```bash
cd upstreams/strix-llama.cpp/tools/server/tests && ./tests.sh unit/test_slot_save.py            # prior art, must keep passing
cd upstreams/strix-llama.cpp/tools/server/tests && ./tests.sh unit/test_nvme_restore_equivalence.py  # planned tracer
curl -s localhost:PORT/metrics | grep llamacpp:prompt_tokens_total
curl -s -X POST localhost:PORT/slots/1 -d '{"action":"save","filename":"a.bin"}'
curl -s -X POST localhost:PORT/slots/0 -d '{"action":"restore","filename":"a.bin"}'
```

## Provenance

`agent://NvmeTestModel` (fixtures/commands; payload JSON-encoded), `agent://NvmeServerAudit` (counters + test specs §8), `agent://NvmeStateAudit` (flags). Prior art paths above. Parent: #59.

**Comment by dymoo** [comment]
**Harness scope complete — closing with evidence; runtime GREEN fix deliberately stays open under #65/#66 (verification run 2026-09-22).**

- Harness at the real-server seam: `scripts/state_restore_harness.py` — uninterrupted HOT arm vs save → erase/free → RESTORED arm with **byte-identical continuation request bytes** (prompts as token-id arrays via `/tokenize`, `add_special:false`); greedy temperature 0 seed 42; per-request `timings.prompt_n`/`cache_n` **and** independent Prometheus `llamacpp:prompt_tokens_total`/`…_cached_total` delta assertions; positions + top-logprob parity; cold probe proves `--cache-ram 0` took effect; recipe checks enforced (`--metrics`, `--slots`, `--slot-save-path`, `--model`+`--offline`, `--threads 1`+`--n-gpu-layers 0`, `--parallel 1`, **no `--swa-full`**); fixture pinned `ggml-org/tinygemma3-GGUF` @ `c287502cd9e278dac8eed805c112cce5d0081e0b`, 47,227,552 bytes, sha256-verified, served offline; never spawns a server itself (external endpoint only). Stdlib only.
- Consumer test: `tests/test_nvme_continuation.py::NvmeContinuationTest::test_hot_vs_restored_continuation_all_assertions_pass` — skips **only** when `NVME_CONTINUATION_BASE_URL` is unset (no other skip condition).
- Full-prefix recomputation cannot pass: `*.prefill_suffix_bound` (named constants: suffix + batch + n_predict + boundary), `compare.prompt_n_parity`, `*.metrics_delta_physical` (global counter delta == `prompt_n`), `*.position_identity` (`prompt_n + cache_n == len(prompt)`) — asserted even when HTTP 200 is returned.
- **RED proof captured** (`docs/nvme-continuation.md`, live endpoint, variant small): `restored.prefill_suffix_bound` observed `prompt_n` **4360** vs bound **777**; `compare.prompt_n_parity` restored **4360** vs hot **257**; `compare.logprob_within_tol` max |Δlogprob| ≈ **0.0266** > tol 0.001 — mechanism cited to pinned `server-context.cpp:2874 / :3617 / :3648-3651`. Save/erase bookkeeping assertions stay green in RED.
- Default-SWA skip situation: the upstream skipped case is replaced by a **tracked failing-first regression** (the RED table above) — acceptance item 5 satisfied as specified ("real assertion or a tracked failing-first regression").
- **Verifying commands:** `python3 scripts/state_restore_harness.py run --base-url http://127.0.0.1:8080 --variant small --report /tmp/nvme-small.json` (exit 1 in RED, failures listed in report) and `NVME_CONTINUATION_BASE_URL=http://127.0.0.1:8080 python3 -m unittest tests.test_nvme_continuation -v`. This session, without a live endpoint: `PYTHONPATH=src python3 -m unittest tests.test_nvme_continuation -v` → **Ran 1 test … OK (skipped=1)** — the documented env gate, as designed.
- Out of scope for this close (kept OPEN): runtime-side equivalence passing + first-class metrics record (`#65`, `#66`), with the fix patch in flight (`patches/strix-slot-restore-checkpoints.patch`, sibling-owned).
